### PR TITLE
Multi-stream JDBC changes

### PR DESCRIPTION
**Pull request by kdeschle**
### What changes were proposed in this pull request?

Performance improvements to the Hive JDBC driver to allow for high performance fetching.

HIVE-28018: Don't require HiveConf for JDBC Metadata calls [Kurt Deschler]
HIVE-27886: Configure Hadoop security for Kerberos when required [Michael Smith]
HIVE-27887: Provide reasonable defaults for ResultSetMetaData [Michael Smith]
HIVE-27873: Fix getOperationStatus() handling and optimize result fetch path [Kurt Deschler]
HIVE-27872: Support multi-stream parallel fetch in JDBC driver [Kurt Deschler]
HIVE-27874: Support datatype conversion on fetch threads [Kurt Deschler]

### Why are the changes needed?

JDBC performance with large datasets was very slow, especially when connected to Impala via HS2 protocol

### Does this PR introduce _any_ user-facing change?

-new JDBC connection parameter 'fetchThreads'
-fetchSize is enforce for prepared statements
-JDBC standalone JAR can be dynamically loaded
-custom hiveconf is no longer required to use the JDBC jar

### Is the change a dependency upgrade?

No

### How was this patch tested?

Ran existing JDBC regression tests
Manual testing multi-stream fetch against Impala

### CODE REVIEW

File: jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java [hunk 1]
```diff
@@ -285,6 +289,16 @@ public HiveConnection(String uri, Properties info) throws SQLException {
     this(uri, info, HiveJdbcBrowserClientFactory.get());
   }
 
+  public HiveConnection Clone() throws SQLException {
```
Comment (aturoczy) [inline jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java:292]: In which use case the Clone is needed? It is not better just to create another connection with the same connection string?

File: jdbc/src/java/org/apache/hive/jdbc/HiveResultSetMetaData.java [hunk 2]
```diff
@@ -30,6 +30,8 @@
  *
  */
 public class HiveResultSetMetaData implements java.sql.ResultSetMetaData {
+  private static final String DOT = ".";
```
Comment (aturoczy) [inline jdbc/src/java/org/apache/hive/jdbc/HiveResultSetMetaData.java:33]: It is a bit overloaded to me to create a constant for DOT that is .  - I'm fine with the  . :-) 

cc: @ayushtkn @deniskuzZ Is this a design rule to make this static final? Even in the simplest cases?

**Comment by aturoczy** [comment]
The test failed with incorrect POM :(

File: jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java (hunk 1 above)
Comment (aturoczy) [inline jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java:292]: Alos the clone is small c :)

File: jdbc/pom.xml [hunk 3]
```diff
@@ -302,6 +295,8 @@
                       <exclude>org.apache.hadoop:hadoop-annotations</exclude>
                       <exclude>org.apache.hadoop:hadoop-auth</exclude>
                       <exclude>org.apache.hbase*:*</exclude>
+                      <exclude>org.apache.hive:hive-classification</exclude>  
+                      <exclude>org.apache.hive:hive-storage-api</exclude>
```
Comment (MikaelSmith) [inline jdbc/pom.xml:299]: These have extra trailing whitespace.

```suggestion
                      <exclude>org.apache.hive:hive-classification</exclude>
                      <exclude>org.apache.hive:hive-storage-api</exclude>
```

File: jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java (hunk 1 above)
Comment (MikaelSmith) [reply jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java:292]: It ensures both clients share a session ID.

Comment (MikaelSmith) [reply jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java]: I'm not sure overriding Object#clone makes sense. I've added a comment to describe it's behavior.

File: jdbc/src/java/org/apache/hive/jdbc/HiveResultSetMetaData.java (hunk 2 above)
Comment (deniskuzZ) [reply jdbc/src/java/org/apache/hive/jdbc/HiveResultSetMetaData.java:33]: all constants should be marked as `static final`

File: jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java (hunk 1 above)
Comment (deniskuzZ) [reply jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java:295]: if you don't won't to override Object#clone, create a copy constructor

File: jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java [hunk 4]
```diff
@@ -307,6 +333,12 @@ protected HiveConnection(String uri, Properties info,
     sessConfMap = connParams.getSessionVars();
     setupLoginTimeout();
     if (isKerberosAuthMode()) {
+      // Ensure UserGroupInformation includes any authorized Kerberos principals.
+      LOG.debug("Configuring Kerberos mode");
+      Configuration config = new Configuration();
+      config.set("hadoop.security.authentication", "Kerberos");
```
Comment (deniskuzZ) [inline jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java:339]: use `AuthTypes.KERBEROS.getAuthName`

File: jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java [hunk 5]
```diff
@@ -293,12 +311,20 @@ protected int getNumRetries() {
   @VisibleForTesting
   protected HiveConnection(String uri, Properties info,
       IJdbcBrowserClientFactory browserClientFactory) throws SQLException {
+    this(uri, info, browserClientFactory, true);
+  }
+
+  @VisibleForTesting
```
Comment (deniskuzZ) [inline jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java:317]: why do we need this annotation, I don't see use in tests for the new constructor

File: jdbc/src/java/org/apache/hive/jdbc/Utils.java [hunk 6]
```diff
@@ -166,6 +166,7 @@ public static class JdbcConnectionParams {
     // Set the fetchSize
     static final String FETCH_SIZE = "fetchSize";
     static final String INIT_FILE = "initFile";
+    static final String FETCH_THREADS = "fetchThreads";
```
Comment (deniskuzZ) [inline jdbc/src/java/org/apache/hive/jdbc/Utils.java:169]: FETCH_THREADS_COUNT = "fetchThreadCount";

File: jdbc/src/java/org/apache/hive/jdbc/JdbcColumn.java [hunk 7]
```diff
@@ -372,6 +372,20 @@ static int columnScale(Type hiveType, JdbcColumnAttributes columnAttributes)
     }
   }
 
+  static boolean columnSigned(Type hiveType, JdbcColumnAttributes columnAttributes)
+      throws SQLException {
+    int columnType = hiveTypeToSqlType(hiveType);
+    switch(columnType) {
+    case Types.TINYINT:
```
Comment (deniskuzZ) [inline jdbc/src/java/org/apache/hive/jdbc/JdbcColumn.java:379]: please format the code

File: jdbc/src/java/org/apache/hive/jdbc/HiveStatement.java [hunk 8]
```diff
@@ -73,12 +74,16 @@ public class HiveStatement implements java.sql.Statement {
   private static final int DEFAULT_FETCH_SIZE =
       HiveConf.ConfVars.HIVE_SERVER2_THRIFT_RESULTSET_DEFAULT_FETCH_SIZE.defaultIntVal;
 
+// TODO: Should there be a hiveconf default?
+  private static final int DEFAULT_FETCH_THREADS = 1;
```
Comment (deniskuzZ) [inline jdbc/src/java/org/apache/hive/jdbc/HiveStatement.java:78]: i think so

File: jdbc/src/java/org/apache/hive/jdbc/HiveResultSetMetaData.java [hunk 9]
```diff
@@ -43,7 +45,9 @@ public HiveResultSetMetaData(List<String> columnNames,
   }
 
   public String getCatalogName(int column) throws SQLException {
-    throw new SQLFeatureNotSupportedException("Method not supported");
+    // Hive doesn't implement a catalog concept, see
+    // https://issues.apache.org/jira/browse/HIVE-3121.
+    return "";
```
Comment (deniskuzZ) [inline jdbc/src/java/org/apache/hive/jdbc/HiveResultSetMetaData.java:50]: should be `default`

File: jdbc/src/java/org/apache/hive/jdbc/HiveResultSetMetaData.java [hunk 10]
```diff
@@ -95,11 +99,23 @@ public int getScale(int column) throws SQLException {
...
+      return tableName.substring(0, index);
+    }
+    // Impala usually doesn't return fully qualified column names. Return "" to avoid
+    // giving false results.
+    return "";
   }
 
   public String getTableName(int column) throws SQLException {
-    throw new SQLFeatureNotSupportedException("Method not supported");
+    String columnName = getColumnName(column);
```
Comment (deniskuzZ) [inline jdbc/src/java/org/apache/hive/jdbc/HiveResultSetMetaData.java:113]: same, use `HiveTableName.of(tableName).getTable()`

File: jdbc/src/java/org/apache/hive/jdbc/ConvertedResultSet.java [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+
+/**
+ * ConvertedResultSet
+ */
+public class ConvertedResultSet implements RowSet {
+
+  private final ArrayList<Object[]> rows;
+  private final int numColumns;
+
+  private interface ConvertFunc {
```
Comment (deniskuzZ) [inline jdbc/src/java/org/apache/hive/jdbc/ConvertedResultSet.java:47]: why not use `UnaryOperator`?

File: jdbc/src/java/org/apache/hive/jdbc/ConvertedResultSet.java [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+    Object convert(Object o);
+  }
+
+  static final int nTypes = Type.values().length;
+  static final ConvertFunc convertFuncs [] = new ConvertFunc [nTypes];
```
Comment (deniskuzZ) [inline jdbc/src/java/org/apache/hive/jdbc/ConvertedResultSet.java:52]: why not use Map?
````
  static final Map<Type, UnaryOperator<?>> convertFuncs = new HashMap<Type, UnaryOperator<?>>() {{
      put(Type.BINARY_TYPE, o -> (o instanceof String) ? ((String) o).getBytes() : 0);
      ....
    }};
````

File: jdbc/src/java/org/apache/hive/jdbc/ConvertedResultSet.java [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+import org.apache.hive.service.cli.RowSet;
+import org.apache.hive.service.cli.TableSchema;
+import org.apache.hive.service.cli.TypeDescriptor;
+
+/**
+ * ConvertedResultSet
+ */
+public class ConvertedResultSet implements RowSet {
+
+  private final ArrayList<Object[]> rows;
```
Comment (deniskuzZ) [inline jdbc/src/java/org/apache/hive/jdbc/ConvertedResultSet.java:44]: please use interface (code to interface, not an implementation)

File: jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java [hunk 14]
```diff
@@ -360,8 +393,10 @@ protected HiveConnection(String uri, Properties info,
         throw new SQLException(new IllegalArgumentException(
             "Browser mode is not supported in embedded mode"));
       }
-      openSession();
-      executeInitSql();
+      if (initSession) {
```
Comment (deniskuzZ) [inline jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java:396]: could we extract this into a helper method?

File: jdbc/src/java/org/apache/hive/jdbc/HiveQueryResultSet.java [hunk 15]
```diff
@@ -98,6 +109,7 @@ public static class Builder {
     private List<String> colTypes;
     private List<JdbcColumnAttributes> colAttributes;
     private int fetchSize = 50;
+    private int fetchThreads = 1;
```
Comment (deniskuzZ) [inline jdbc/src/java/org/apache/hive/jdbc/HiveQueryResultSet.java:112]: should we get this from HiveConf?

File: jdbc/src/java/org/apache/hive/jdbc/HiveQueryResultSet.java [hunk 16]
```diff
@@ -271,8 +299,18 @@ private void setSchema(List<String> colNames, List<String> colTypes,
...
+      fetchedRows = RowSetFactory.create(new TRowSet(), protocol);
+      fetchedRowsItr = fetchedRows.iterator();
+    } catch (TException e) {
+      throw new SQLException(e);
+    }
+  }
+
   @Override
   public void close() throws SQLException {
+    ShutdownPool();
```
Comment (deniskuzZ) [inline jdbc/src/java/org/apache/hive/jdbc/HiveQueryResultSet.java:313]: please use `camel-case` namings

File: jdbc/src/java/org/apache/hive/jdbc/HiveQueryResultSet.java [hunk 17]
```diff
@@ -290,7 +328,7 @@ public void close() throws SQLException {
     client = null;
     stmtHandle = null;
     isClosed = true;
-    operationStatus = null;
+    InitEmptyIterator();
```
Comment (deniskuzZ) [inline jdbc/src/java/org/apache/hive/jdbc/HiveQueryResultSet.java:331]: `camel-case`

File: jdbc/src/java/org/apache/hive/jdbc/HiveQueryResultSet.java [hunk 18]
```diff
@@ -307,66 +345,220 @@ private void closeOperationHandle(TOperationHandle stmtHandle) throws SQLExcepti
...
 
-  /**
-   * Moves the cursor down one row from its current position.
-   *
-   * @see java.sql.ResultSet#next()
-   * @throws SQLException
-   *           if a database access error occurs.
-   */
-  public boolean next() throws SQLException {
+  ExecutorService pool = null;
```
Comment (deniskuzZ) [inline jdbc/src/java/org/apache/hive/jdbc/HiveQueryResultSet.java:348]: could you please move it to the rest or class fields declarations

File: jdbc/src/java/org/apache/hive/jdbc/HiveQueryResultSet.java [hunk 19]
```diff
@@ -307,66 +345,220 @@ private void closeOperationHandle(TOperationHandle stmtHandle) throws SQLExcepti
...
+          gotLastBatch = true;
+        }
+        fetchedRows = RowSetFactory.create(results, protocol);
+        if (useConvertedResultSet) {
+          fetchedRows = new ConvertedResultSet(fetchedRows, getSchema());
+        }
+
+        nextStartRow.set(results.getStartRowOffset() + 1 + fetchedRows.numRows());
+      } catch (TException ex) {
+        ex.printStackTrace();
```
Comment (deniskuzZ) [inline jdbc/src/java/org/apache/hive/jdbc/HiveQueryResultSet.java:408]: use `Logger` instead

File: jdbc/src/java/org/apache/hive/jdbc/HiveQueryResultSet.java [hunk 20]
```diff
@@ -307,66 +345,220 @@ private void closeOperationHandle(TOperationHandle stmtHandle) throws SQLExcepti
...
+        throw new SQLException("Error retrieving next row", ex);
+      }
+    } else {
+      if (!gotLastBatch) {
+        if (pool == null) {
+          pool = Executors.newFixedThreadPool(poolSize);
+        }
+        // Add another thread on each row batch up to the limit
+        if(threadsStarted < (hasStartRow ? poolSize : 1)) {
```
Comment (deniskuzZ) [inline jdbc/src/java/org/apache/hive/jdbc/HiveQueryResultSet.java:417]: nit: space

File: jdbc/src/java/org/apache/hive/jdbc/HiveQueryResultSet.java [hunk 21]
```diff
@@ -307,66 +345,220 @@ private void closeOperationHandle(TOperationHandle stmtHandle) throws SQLExcepti
...
+          final boolean useMainClient = (threadsStarted == 0);
+          threadsStarted++;
+          pool.execute(()->{
```
Comment (deniskuzZ) [inline jdbc/src/java/org/apache/hive/jdbc/HiveQueryResultSet.java:420]: nit: space after `->`

File: jdbc/src/java/org/apache/hive/jdbc/HiveQueryResultSet.java [hunk 22]
```diff
@@ -307,66 +345,220 @@ private void closeOperationHandle(TOperationHandle stmtHandle) throws SQLExcepti
...
+            }
+            // TODO: shutdown called here can make execute return
+            // RejectedExecutionException
+            // pool.shutdownNow();
+          });
+        }
+      }
+      try {
+        if (interruptException != null) {
+          throw interruptException;
```
Comment (deniskuzZ) [inline jdbc/src/java/org/apache/hive/jdbc/HiveQueryResultSet.java:500]: show we `shutdown` the pool?

File: jdbc/src/java/org/apache/hive/jdbc/HiveQueryResultSet.java [hunk 23]
```diff
@@ -307,66 +345,220 @@ private void closeOperationHandle(TOperationHandle stmtHandle) throws SQLExcepti
...
+        }
+        FetchResult result = resultQueue.take();
+        fetchDone = !result.hasMoreRows;
+        if (result.ex != null) {
+          ShutdownPool();
+          throw new SQLException(result.ex);
+        }
+        fetchedRows = result.fetchedRows;
+      } catch (InterruptedException e) {
```
Comment (deniskuzZ) [inline jdbc/src/java/org/apache/hive/jdbc/HiveQueryResultSet.java:509]: should we `shutdown` the pool in `finally` block?

File: jdbc/src/java/org/apache/hive/jdbc/HiveQueryResultSet.java [hunk 24]
```diff
@@ -307,66 +345,220 @@ private void closeOperationHandle(TOperationHandle stmtHandle) throws SQLExcepti
...
 
-      if (!fetchedRowsItr.hasNext()) {
-        return false;
+    orientation = TFetchOrientation.FETCH_NEXT;
+    fetchedRowsItr = fetchedRows.iterator();
+
+    return fetchedRowsItr.hasNext();
+  }
+
+  void DrainQueue() throws SQLException {
```
Comment (deniskuzZ) [inline jdbc/src/java/org/apache/hive/jdbc/HiveQueryResultSet.java:520]: `camel-case`

File: jdbc/src/java/org/apache/hive/jdbc/HiveQueryResultSet.java [hunk 25]
```diff
@@ -307,66 +345,220 @@ private void closeOperationHandle(TOperationHandle stmtHandle) throws SQLExcepti
...
+    }
+  }
 
-      row = fetchedRowsItr.next();
-      rowsFetched++;
-    } catch (SQLException eS) {
-      throw eS;
-    } catch (Exception ex) {
-      throw new SQLException("Error retrieving next row", ex);
+  void ShutdownPool() throws SQLException {
```
Comment (deniskuzZ) [inline jdbc/src/java/org/apache/hive/jdbc/HiveQueryResultSet.java:529]: `camel-case`

File: jdbc/src/java/org/apache/hive/jdbc/HiveQueryResultSet.java [hunk 26]
```diff
@@ -307,66 +345,220 @@ private void closeOperationHandle(TOperationHandle stmtHandle) throws SQLExcepti
...
+    if (pool != null) {
+      poolDone = true;
+      DrainQueue();
```
Comment (deniskuzZ) [inline jdbc/src/java/org/apache/hive/jdbc/HiveQueryResultSet.java:532]: do we need to call it before pool shutdown?

File: jdbc/src/java/org/apache/hive/jdbc/HiveQueryResultSet.java [hunk 27]
```diff
@@ -307,66 +345,220 @@ private void closeOperationHandle(TOperationHandle stmtHandle) throws SQLExcepti
...
+      pool.shutdownNow();
+      try {
+        while (!pool.awaitTermination(1, TimeUnit.SECONDS)) {
+          DrainQueue();
```
Comment (deniskuzZ) [inline jdbc/src/java/org/apache/hive/jdbc/HiveQueryResultSet.java:536]: why do you need to call it multiple times?

**Comment by deniskuzZ** [comment]
thanks for the PR @MikaelSmith , @kdeschle! In general LGTM, left few comments

File: jdbc/src/java/org/apache/hive/jdbc/ConvertedResultSet.java (hunk 11 above)
Comment (kdeschle) [reply jdbc/src/java/org/apache/hive/jdbc/ConvertedResultSet.java:47]: That leads to generic array creation compile error.

File: jdbc/src/java/org/apache/hive/jdbc/ConvertedResultSet.java (hunk 12 above)
Comment (kdeschle) [reply jdbc/src/java/org/apache/hive/jdbc/ConvertedResultSet.java:52]: Map would be less efficient than the current ordinal lookup.

File: jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java (hunk 14 above)
Comment (kdeschle) [reply jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java:396]: I don't think it makes sense to add a function here.

File: jdbc/src/java/org/apache/hive/jdbc/HiveQueryResultSet.java (hunk 23 above)
Comment (kdeschle) [reply jdbc/src/java/org/apache/hive/jdbc/HiveQueryResultSet.java:509]: Not here. This is just early shutdown for error case. Unconditional shutdown is from close()

File: jdbc/src/java/org/apache/hive/jdbc/HiveQueryResultSet.java (hunk 26 above)
Comment (kdeschle) [reply jdbc/src/java/org/apache/hive/jdbc/HiveQueryResultSet.java:532]: This sequence is necessary to ensure that nothing blocks and nothing is left in the queue.

File: jdbc/src/java/org/apache/hive/jdbc/HiveQueryResultSet.java (hunk 27 above)
Comment (kdeschle) [reply jdbc/src/java/org/apache/hive/jdbc/HiveQueryResultSet.java:536]: This was part of prescribed strategies for shutting down ExecutorService.

File: jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java (hunk 1 above)
Comment (kdeschle) [reply jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java:295]: switched to copy constructor

File: jdbc/src/java/org/apache/hive/jdbc/HiveQueryResultSet.java (hunk 17 above)
Comment (kdeschle) [reply jdbc/src/java/org/apache/hive/jdbc/HiveQueryResultSet.java:331]: Have to restore initial state in case result is re-opened.

**Comment by deniskuzZ** [comment]
hi @kdeschle, thanks for addressing the review comments! 
Would you be able to rebase this PR?

File: jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java (hunk 4 above)
Comment (kdeschle) [reply jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java:339]: changed constant.
I don't thing it should  always be kerberized.

File: jdbc/src/java/org/apache/hive/jdbc/ConvertedResultSet.java (hunk 12 above)
Comment (deniskuzZ) [reply jdbc/src/java/org/apache/hive/jdbc/ConvertedResultSet.java:52]: That is the usage pattern for the HashTable - do the lookups. I believe this code part does not need to be optimized at the expense of readability.
````
  static final Map<Type, UnaryOperator<Object>> convertFuncs = new HashMap<Type, UnaryOperator<Object>>() {{
    put(Type.BINARY_TYPE, o -> (o instanceof String) ? ((String) o).getBytes() : 0);
    put(Type.TIMESTAMP_TYPE, o -> Timestamp.valueOf((String) o));
    put(Type.TIMESTAMPLOCALTZ_TYPE, o -> TimestampTZUtil.parse((String) o));
    put(Type.DECIMAL_TYPE, o -> new BigDecimal((String) o));
    put(Type.DATE_TYPE, o -> Date.valueOf((String) o));
    put(Type.INTERVAL_YEAR_MONTH_TYPE, o -> HiveIntervalYearMonth.valueOf((String) o));
    put(Type.INTERVAL_DAY_TIME_TYPE, o -> HiveIntervalDayTime.valueOf((String) o));
  }};

  public ConvertedResultSet(RowSet rowSet, TableSchema schema) {
    rows = new ArrayList<>();
    numColumns = rowSet.numColumns();
    Iterator<Object[]> srcIter = rowSet.iterator();
    final int nCols = schema.getSize();
    
    List<UnaryOperator<Object>> colConvert = new ArrayList<>(nCols);
    for (int i = 0; i < nCols; ++i) {
      colConvert.add(convertFuncs.get(schema.getColumnDescriptorAt(i).getType()));
    }
    
    while (srcIter.hasNext()) {
      Object[] srcRow = srcIter.next();
      Object[] dstRow = new Object[srcRow.length];
      for (int i = 0; i < nCols; ++i) {
        if (colConvert.get(i) != null && srcRow[i] != null) {
          dstRow[i] = colConvert.get(i).apply(srcRow[i]);
        } else {
          dstRow[i] = srcRow[i];
        }
      }
      rows.add(dstRow);
    }
  }
````

File: jdbc/src/java/org/apache/hive/jdbc/ConvertedResultSet.java [hunk 28]
```diff
@@ -0,0 +1,151 @@
...
+public class ConvertedResultSet implements RowSet {
+
+  private final List<Object[]> rows;
+  private final int numColumns;
+
+  private interface ConvertFunc {
+    Object convert(Object o);
+  }
+
+  static final int nTypes = Type.values().length;
```
Comment (deniskuzZ) [inline jdbc/src/java/org/apache/hive/jdbc/ConvertedResultSet.java:51]: `nTypes` variable can be removed

File: jdbc/src/java/org/apache/hive/jdbc/ConvertedResultSet.java (hunk 11 above)
Comment (deniskuzZ) [reply jdbc/src/java/org/apache/hive/jdbc/ConvertedResultSet.java:47]: it doesn't, please see snippet in `convertFuncs` comment

File: jdbc/src/java/org/apache/hive/jdbc/ConvertedResultSet.java [hunk 29]
```diff
@@ -0,0 +1,151 @@
...
+        return iterator.hasNext();
+      }
+
+      @Override
+      public Object[] next() {
+        return iterator.next();
+      }
+
+      @Override
+      public void remove() {
```
Comment (deniskuzZ) [inline jdbc/src/java/org/apache/hive/jdbc/ConvertedResultSet.java:146]: no need to override, it's identical to parent method

**Comment by deniskuzZ** [comment]
In general LGTM. Would be great to include this perf improvement in the next 4.1.0 Hive release. 

I've reopened 1 comment with the refactor recommendation:  https://github.com/apache/hive/pull/4902#discussion_r1886898843

Also, Sonar reported a few new issues, we should try to address those.

File: jdbc/src/java/org/apache/hive/jdbc/ConvertedResultSet.java (hunk 12 above)
Comment (kdeschle) [reply jdbc/src/java/org/apache/hive/jdbc/ConvertedResultSet.java:52]: With this change there was a 1.1% decrease in performance. I can keep the change if you insist.

Comment (deniskuzZ) [reply jdbc/src/java/org/apache/hive/jdbc/ConvertedResultSet.java:52]: @kdeschle what Map implementation have you used? EnumMap to HashMap?
EnumMap is 1-to-1 match with your solution (vals array with key.ordinal() index access)

1.1% decrease in generic performance performance or collection access?

Comment (kdeschle) [reply jdbc/src/java/org/apache/hive/jdbc/ConvertedResultSet.java]: Nevermind. It seems it was just variation in my runs. Posted new rev with these changes and some fixes.

**Review by deniskuzZ — APPROVED** [review]
+1, pending green build

**Comment by deniskuzZ** [comment]
Sonar reported a number of newly introduced unused import, methods and vars, could we please drop them:
1. https://sonarcloud.io/project/issues?sinceLeakPeriod=true&issueStatuses=OPEN%2CCONFIRMED&pullRequest=4902&id=apache_hive&open=AZPVJuwuQje9E2BL3_-t
2. https://sonarcloud.io/project/issues?sinceLeakPeriod=true&issueStatuses=OPEN%2CCONFIRMED&pullRequest=4902&id=apache_hive&open=AZPFqXK8BQZ6BS4X3Vch
3. https://sonarcloud.io/project/issues?sinceLeakPeriod=true&issueStatuses=OPEN%2CCONFIRMED&pullRequest=4902&id=apache_hive&open=AZPFqXNlBQZ6BS4X3Vc4
4. https://sonarcloud.io/project/issues?sinceLeakPeriod=true&issueStatuses=OPEN%2CCONFIRMED&pullRequest=4902&id=apache_hive&open=AZPFqXNlBQZ6BS4X3Vc5
5. https://sonarcloud.io/project/issues?sinceLeakPeriod=true&issueStatuses=OPEN%2CCONFIRMED&pullRequest=4902&id=apache_hive&open=AZPFqXNlBQZ6BS4X3Vcv
6. https://sonarcloud.io/project/issues?sinceLeakPeriod=true&issueStatuses=OPEN%2CCONFIRMED&pullRequest=4902&id=apache_hive&open=AZPFqXNlBQZ6BS4X3Vcw
7. https://sonarcloud.io/project/issues?sinceLeakPeriod=true&issueStatuses=OPEN%2CCONFIRMED&pullRequest=4902&id=apache_hive&open=AZPFqXHnBQZ6BS4X3Vcg

**Review by aturoczy — APPROVED** [review]

**Review by MikaelSmith — APPROVED** [review]

**Comment by deniskuzZ** [comment]
this PR causes issues on a Jenkins, build timeouts after 12hr

**Comment by deniskuzZ** [comment]
i checked the test bundle and there are no failures. rebased with master

**Comment by kdeschle** [comment]
Any idea what the build timeout was about?

**Comment by deniskuzZ** [comment]
it seems that all splits are done except split-13. It's already more than 5hr

**Comment by kdeschle** [comment]
Thanks for the insights. I was able to repro locally as well and fix the issues.

**Comment by deniskuzZ** [comment]
> Thanks for the insights. I was able to repro locally as well and fix the issues.

Thanks for the fix @kdeschle!

I noticed that we have 2 commits for HIVE-27872, should I squash them into 1
1. Support multi-stream parallel fetch in JDBC driver
2. Don't require HiveConf for JDBC Metadata calls 

Also [HIVE-27882](https://issues.apache.org/jira/browse/HIVE-27882): Fix missing classes in standalone jar, mentioned in the PR description, is missing. Is that expected?

**Comment by kdeschle** [comment]
Closed HIVE-27882 as dup of HIVE-28315 and remove from PR desc since that was already fixed.
Updated the other ticket to HIVE-28018.

I'd like to leave the stack as-is now as they address different problems and can be cherrypicked separately to other branches.

**Comment by deniskuzZ** [comment]
merged to master
thanks @MikaelSmith  & @kdeschle for the contribution!

**Comment by kdeschle** [comment]
Thanks Denis!

**Pull request closed without merging**